src/poprox_recommender/embedders/article.py

```diff
@@ -52,28 +58,33 @@ def __call__(self, article_set: ArticleSet) -> ArticleSet:
                             article.title, padding="max_length", max_length=TITLE_LENGTH_LIMIT, truncation=True
                         ),
                         dtype=th.int32,
-                    )
+                    ).to(self.device)
                     for article in uncached
                 ]
             )
-            assert_tensor_size(uc_title_tokens, len(uncached), TITLE_LENGTH_LIMIT)
+            assert_tensor_size(uc_title_tokens, len(uncached), TITLE_LENGTH_LIMIT, label="uncached title tokens")
 
             # Step 4: embed the uncached articles
             uc_embeddings = self.model.get_news_vector(uc_title_tokens)
-            assert uc_embeddings.shape[0] == len(uncached)
+            assert_tensor_size(
+                uc_embeddings, len(uncached), self.model.embedding_size, label="uncached article embeddings"
+            )
 
             # Step 5: store embeddings to cache & result
             for i, uca in enumerate(uncached):
                 # copy the tensor so it isn't attached to excess memory
                 a_emb = uc_embeddings[i, :].clone()
                 cached[uca.article_id] = a_emb
-                self.embedding_cache[uca.article_id] = a_emb
+                # use CPU for cached articles to reduce CUDA memory usage
+                self.embedding_cache[uca.article_id] = a_emb.cpu()
 
         # Step 6: stack the embeddings into a single tensor
-        embed_single_tensors = list(cached.values())
-        assert not any(e is None for e in embed_single_tensors)
+        # we do this with a list to properly deal with duplicate articles
+        embed_single_tensors = [cached[article.article_id].to(self.device) for article in article_set.articles]  # type: ignore
```

Comment:
Our MIND data can have duplicate articles, so this makes the embeddings match the article set's list even when there are duplicates.